pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1809,3 +1986,146 @@ func getSidecarInboundBindIP(node *model.Proxy) string {
+				// This is a catch all filter chain.
+				// We can only merge with a non-catch all filter chain
+				// Else mark it as conflict
+				if incomingFilterChain.FilterChainMatch == nil {
+					// NOTE: While pluginParams.Service can be nil,
+					// this code cannot be reached if Service is nil because a pluginParams.Service can be nil only
+					// for user defined Egress listeners with ports. And these should occur in the API before
+					// the wildcard egress listener. the check for the "locked" bit will eliminate the collision.
+					// User is also not allowed to add duplicate ports in the egress listener
+					var newHostname host.Name
+					if pluginParams.Service != nil {
+						newHostname = pluginParams.Service.Hostname
+					} else {
+						// user defined outbound listener via sidecar API
+						newHostname = "sidecar-config-egress-tcp-listener"
+					}
+
+					conflictFound = true
+					outboundListenerConflict{
+						metric:          model.ProxyStatusConflictOutboundListenerTCPOverTCP,
+						node:            pluginParams.Node,
+						listenerName:    listenerMapKey,
+						currentServices: currentListenerEntry.services,
+						currentProtocol: currentListenerEntry.servicePort.Protocol,
+						newHostname:     newHostname,
+						newProtocol:     pluginParams.Port.Protocol,
+					}.addMetric(node, pluginParams.Push)
+					break compareWithExisting
+				} else {
+					continue
+				}
+			}
+			if incomingFilterChain.FilterChainMatch == nil {
+				continue
+			}
+
+			// We have two non-catch all filter chains. Check for duplicates
+			if reflect.DeepEqual(*existingFilterChain.FilterChainMatch, *incomingFilterChain.FilterChainMatch) {
+				var newHostname host.Name
+				if pluginParams.Service != nil {
+					newHostname = pluginParams.Service.Hostname
+				} else {
+					// user defined outbound listener via sidecar API
+					newHostname = "sidecar-config-egress-tcp-listener"
+				}
+
+				conflictFound = true
+				outboundListenerConflict{
+					metric:          model.ProxyStatusConflictOutboundListenerTCPOverTCP,
+					node:            pluginParams.Node,
+					listenerName:    listenerMapKey,
+					currentServices: currentListenerEntry.services,
+					currentProtocol: currentListenerEntry.servicePort.Protocol,
+					newHostname:     newHostname,
+					newProtocol:     pluginParams.Port.Protocol,
+				}.addMetric(node, pluginParams.Push)
+				break compareWithExisting
+			}
+		}
+
+		if !conflictFound {
+			// There is no conflict with any filter chain in the existing listener.
+			// So append the new filter chains to the existing listener's filter chains
+			newFilterChains = append(newFilterChains, incomingFilterChain)
+			if pluginParams.Service != nil {
+				lEntry := listenerMap[listenerMapKey]
+				lEntry.services = append(lEntry.services, pluginParams.Service)
+			}
+		}
+	}
+
+	return newFilterChains
+}
+
+func mergeFilterChains(httpFilterChain, tcpFilterChain []*listener.FilterChain) []*listener.FilterChain {
```

Comment:
this does not merge filterchainmatch fully. Is this the behavior desired? if it is please make a big warning so others do not use this function and get unexpected results

Reply:
> this does not merge filterchainmatch fully.
The function merges http filter chain with tcp filter chain. Why do they need to merge filter chain match?

Reply:
You are already partially merging the matches

`fc.FilterChainMatch.ApplicationProtocols = append(fc.FilterChainMatch.ApplicationProtocols, []string{"h2", "http/1.1", "http/1.0"}...)`

Which is not at all clear when you look at the callsite and just see `mergeFilterChains`, I would expect it to merge all matches not just applicationprotocols

Reply:
Nevermind I see now.. you are not merging them you are just adding those application protocols to the match.